webui/react/src/components/MetricSelectFilter.tsx

```diff
@@ -62,16 +98,89 @@ const MetricSelectFilter: React.FC<Props> = ({ metricNames, multiple, onChange,
     (onChange as MultipleHandler)(newMetric.sort(metricNameSorter));
   }, [ multiple, onChange, value ]);
 
+  const handleFiltering = useCallback((search: string, option) => {
+    if (option.key === allOptionId || option.key === resetOptionId) {
+      return true;
+    }
+    if (!option.value) {
+      // Handle optionGroups that don't have a value to make TS happy. They aren't
+      // impacted by filtering anyway
+      return false;
+    }
+    const metricName = metricNameFromValue(option.value);
+    if (metricName === undefined) {
+      // Handle metric values that don't start with 'training|' or 'validation|'. This
+      // shouldn't ever happen and metricNameFromValue logs an error if it does.
+      return false;
+    }
+    return filterFn(search, metricName.name);
+  }, []);
+
+  const handleSearchInputChange = (searchInput: string) => {
+    setFilterString(searchInput);
+  };
+
+  const handleBlur = () => {
+    setFilterString('');
+  };
+
+  const allOption = useMemo(() => {
+    let allOptionLabel;
+    const numVisibleOptions = visibleMetrics.length;
+    if (numVisibleOptions === totalNumMetrics) {
+      allOptionLabel = 'All';
+    } else {
+      allOptionLabel =`All ${numVisibleOptions} results`;
+    }
+    return (
+      <Option key={allOptionId} value={allOptionId}>
+        <BadgeTag label={allOptionLabel} />
+      </Option>
+    );
+
+  }, [ totalNumMetrics, visibleMetrics ]);
+
+  const [ maxTagCount, selectorPlaceholder ] = useMemo(() => {
+    // This should never happen, but fall back to inoffensive empty placeholder
+    if (metricValues === undefined) {
+      return [ 0, '' ];
+    }
+    if (metricValues.length === 0) {
+      // If we set maxTagCount=0 in this case, this placeholder will not be displayed.
+      return [ -1, 'None selected' ];
+    } else if (metricValues.length === totalNumMetrics) {
+      // If we set maxTagCount=-1 in these cases, it will display tags instead of the placeholder.
+      return [ 0, `All ${totalNumMetrics} selected` ];
+    } else {
+      return [ 0, `${metricValues.length} of ${totalNumMetrics} selected` ];
+    }
+  }, [ metricValues, totalNumMetrics ]);
+
   return <SelectFilter
+    autoClearSearchValue={false}
     disableTags
     dropdownMatchSelectWidth={400}
+    filterOption={handleFiltering}
     label="Metrics"
+    maxTagCount={maxTagCount}
+    maxTagPlaceholder={selectorPlaceholder}
     mode={multiple ? 'multiple' : undefined}
+    ref={selectRef}
     showArrow
-    style={{ width: 150 }}
+    style={{ width: 200 }}
     value={metricValues}
+    onBlur={handleBlur}
     onDeselect={handleMetricDeselect}
+    onSearch={handleSearchInputChange}
     onSelect={handleMetricSelect}>
+
+    { multiple && visibleMetrics.length > 0 &&
+    <Option key={resetOptionId} value={resetOptionId}>
+      <BadgeTag label='Reset to Default' />
+    </Option>}
+
+    { multiple && visibleMetrics.length > 1 && allOption}
```

Comment:
**question:** Should we make these persist (without a dependency on `visibleMetrics.length`)? Where I found this to be a little confusing is when I was changing the filtering text, it dynamically added and removed the `Rest to Default` and `All` options.